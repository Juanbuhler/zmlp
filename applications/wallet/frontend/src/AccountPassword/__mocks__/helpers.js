/**
 * onSubmit
 */

export const onSubmit = ({ dispatch }) =>
  dispatch({
    currentPassword: 'foo',
    newPassword: 'bar',
    confirmPassword: 'bar',
  })
