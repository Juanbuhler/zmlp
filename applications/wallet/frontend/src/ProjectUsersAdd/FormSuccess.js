import PropTypes from 'prop-types'
import Link from 'next/link'

import { colors, spacing, typography } from '../Styles'

import FormError from '../FormError'
import FormSuccess from '../FormSuccess'
import Button, { VARIANTS } from '../Button'
import ButtonGroup from '../Button/Group'
import SectionTitle from '../SectionTitle'

const ProjectUsersAddFormSuccess = ({
  projectId,
  succeeded,
  failed,
  onReset,
}) => {
  const { permissions } = succeeded[0]

  return (
    <div>
      {failed.length > 0 && (
        <div>
          <FormError>Users Not Added!</FormError>
          <SectionTitle>Users that Need an Account</SectionTitle>
          <div
            css={{
              paddingTop: spacing.moderate,
              paddingBottom: spacing.moderate,
              color: colors.structure.steel,
            }}>
            {failed.map(user => user.email).join(', ')}
          </div>
        </div>
      )}

      <FormSuccess>Users Added!</FormSuccess>

      <SectionTitle>Users Added</SectionTitle>
      <div
        css={{
          paddingTop: spacing.moderate,
          color: colors.structure.steel,
        }}>
        {succeeded.map(user => user.email).join(', ')}
      </div>

      <SectionTitle>Permissions</SectionTitle>
      <ul css={{ color: colors.structure.zinc }}>
        {permissions.map(permission => (
          <li css={{ fontWeight: typography.weight.bold }} key={permission}>
            {permission.replace(/([A-Z])/g, match => ` ${match}`)}
          </li>
        ))}
      </ul>

      <ButtonGroup>
        <Button variant={VARIANTS.SECONDARY} onClick={onReset}>
          Add Another User
        </Button>
        <Link
          href="/[projectId]/users"
          as={'/[projectId]/users'.replace('[projectId]', projectId)}
          passHref>
          <Button variant={VARIANTS.PRIMARY}>View All</Button>
        </Link>
      </ButtonGroup>
    </div>
  )
}

ProjectUsersAddFormSuccess.propTypes = {
  projectId: PropTypes.string.isRequired,
  succeeded: PropTypes.arrayOf(
    PropTypes.shape({
      email: PropTypes.string.isRequired,
      permissions: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    }).isRequired,
  ).isRequired,
  failed: PropTypes.arrayOf(
    PropTypes.shape({
      email: PropTypes.string.isRequired,
      permissions: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    }).isRequired,
  ).isRequired,
  onReset: PropTypes.func.isRequired,
}

export default ProjectUsersAddFormSuccess
