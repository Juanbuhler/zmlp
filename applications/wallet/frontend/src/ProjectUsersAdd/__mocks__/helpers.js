/**
 * onSubmit
 */

export const onSubmit = ({ dispatch }) =>
  dispatch({
    succeeded: [
      {
        email: 'jane@zorroa.com',
        permissions: [
          'SuperAdmin',
          'ProjectAdmin',
          'AssetsRead',
          'AssetsImport',
        ],
      },
    ],
    failed: [
      {
        email: 'joe@zorroa.com',
        permissions: [
          'SuperAdmin',
          'ProjectAdmin',
          'AssetsRead',
          'AssetsImport',
        ],
      },
    ],
  })

/**
 * onCopy
 */

let mockOnCopy = () => {}

export const __setMockOnCopy = fn => {
  mockOnCopy = fn
}

export const onCopy = (...args) => {
  mockOnCopy(...args)
}
