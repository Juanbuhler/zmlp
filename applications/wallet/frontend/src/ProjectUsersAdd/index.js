import Head from 'next/head'

import PageTitle from '../PageTitle'
import Tabs from '../Tabs'

import ProjectUsersAddForm from './Form'

const ProjectUsersAdd = () => {
  return (
    <>
      <Head>
        <title>User Admin</title>
      </Head>

      <PageTitle>Project User Admin</PageTitle>

      <Tabs
        tabs={[
          { title: 'Users', href: '/[projectId]/users' },
          { title: 'Create User', href: '/[projectId]/users/add' },
        ]}
      />

      <ProjectUsersAddForm />
    </>
  )
}

export default ProjectUsersAdd
