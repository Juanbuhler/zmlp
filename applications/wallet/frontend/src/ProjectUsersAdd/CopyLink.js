import { useRef } from 'react'

import { spacing, colors, constants, typography } from '../Styles'

import Button, { VARIANTS } from '../Button'

import { onCopy } from './helpers'

const ProjectUsersAddCopyLink = () => {
  const inputRef = useRef()

  return (
    <div css={{ paddingTop: spacing.moderate }}>
      <label
        htmlFor="copyLink"
        css={{
          display: 'block',
          color: colors.structure.steel,
        }}>
        A user must have an account to be added. Accounts can be created here:
      </label>
      <div
        css={{
          display: 'flex',
          alignItems: 'center',
          paddingTop: spacing.normal,
        }}>
        <input
          ref={inputRef}
          id="copyLink"
          type="text"
          defaultValue={`${window.location.hostname}/create-account`}
          css={{
            width: constants.form.maxWidth,
            fontSize: typography.size.regular,
            lineHeight: typography.height.regular,
            color: colors.structure.white,
            backgroundColor: colors.structure.mattGrey,
            borderRadius: constants.borderRadius.small,
            padding: spacing.moderate,
            border: 'none',
            resize: 'none',
          }}
        />
        <div
          css={{
            display: 'flex',
            alignItems: 'center',
            paddingLeft: spacing.small,
            paddingRight: spacing.small,
          }}>
          <Button
            variant={VARIANTS.LINK}
            style={{
              '&:active': {
                opacity: 0,
                transition: 'opacity .3s ease',
              },
            }}
            onClick={() => onCopy({ inputRef })}>
            Copy Link
          </Button>
        </div>
      </div>
    </div>
  )
}

export default ProjectUsersAddCopyLink
