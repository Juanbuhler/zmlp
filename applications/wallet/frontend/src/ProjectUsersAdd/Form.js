import { useReducer } from 'react'
import { useRouter } from 'next/router'
import useSWR from 'swr'

import Form from '../Form'
import SectionTitle from '../SectionTitle'
import Input, { VARIANTS as INPUT_VARIANTS } from '../Input'
import { VARIANTS as CHECKBOX_VARIANTS } from '../Checkbox'
import CheckboxGroup from '../Checkbox/Group'
import Button, { VARIANTS as BUTTON_VARIANTS } from '../Button'
import ButtonGroup from '../Button/Group'
import Loading from '../Loading'

import { onSubmit } from './helpers'

import ProjectUsersAddFormResponse from './FormResponse'

const INITIAL_STATE = {
  emails: '',
  permissions: {},
  succeeded: [],
  failed: [],
  errors: {},
}

const reducer = (state, action) => ({ ...state, ...action })

const ProjectUsersAddForm = () => {
  const {
    query: { projectId },
  } = useRouter()

  const { data: { results: permissions } = {} } = useSWR(
    `/api/v1/projects/${projectId}/permissions/`,
  )

  const [state, dispatch] = useReducer(reducer, INITIAL_STATE)

  if (!Array.isArray(permissions)) return <Loading />

  if (state.succeeded.length > 0 || state.failed.length > 0) {
    return (
      <ProjectUsersAddFormResponse
        projectId={projectId}
        succeeded={state.succeeded}
        failed={state.failed}
        permissions={state.permissions}
        onReset={() => dispatch(INITIAL_STATE)}
      />
    )
  }

  return (
    <Form>
      <SectionTitle>Invite User to view projects</SectionTitle>

      <Input
        autoFocus
        id="emails"
        variant={INPUT_VARIANTS.SECONDARY}
        label="Email(s)"
        type="text"
        value={state.emails}
        onChange={({ target: { value } }) => dispatch({ emails: value })}
        hasError={state.errors.name !== undefined}
        errorMessage={state.errors.name}
      />

      <CheckboxGroup
        legend="Add Permissions"
        onClick={permission =>
          dispatch({ permissions: { ...state.permissions, ...permission } })
        }
        options={permissions.map(({ name, description }) => ({
          key: name,
          label: name.replace(/([A-Z])/g, match => ` ${match}`),
          icon: '',
          legend: description,
          initialValue: false,
        }))}
        variant={CHECKBOX_VARIANTS.PRIMARY}
      />

      <ButtonGroup>
        <Button
          type="submit"
          variant={BUTTON_VARIANTS.PRIMARY}
          onClick={() => onSubmit({ projectId, dispatch, state })}
          isDisabled={!state.emails}>
          Send Invite
        </Button>
      </ButtonGroup>
    </Form>
  )
}

export default ProjectUsersAddForm
