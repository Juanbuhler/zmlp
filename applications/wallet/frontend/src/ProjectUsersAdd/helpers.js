import { fetcher } from '../Fetch/helpers'

export const onSubmit = async ({
  projectId,
  dispatch,
  state: { emails: e, permissions: p },
}) => {
  try {
    const emails = e.split(',').map(str => str.trim(''))
    const permissions = Object.keys(p).filter(name => p[name])
    const body = JSON.stringify({
      batch: emails.map(email => ({ email, permissions })),
    })

    const {
      results: { succeeded, failed },
    } = await fetcher(`/api/v1/projects/${projectId}/users/`, {
      method: 'POST',
      body,
    })

    dispatch({
      succeeded: succeeded.map(user => ({
        email: user.email,
        permissions: user.permissions,
      })),
      failed: failed.map(user => ({
        email: user.email,
        permissions: user.permissions,
      })),
    })
  } catch (response) {
    const errors = await response.json()
    const parsedErrors = Object.keys(errors).reduce((acc, errorKey) => {
      acc[errorKey] = errors[errorKey].join(' ')

      return acc
    }, {})

    dispatch({ errors: parsedErrors })
  }
}
