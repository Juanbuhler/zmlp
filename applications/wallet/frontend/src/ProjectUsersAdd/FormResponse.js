import PropTypes from 'prop-types'
import Link from 'next/link'

import { colors, spacing, typography } from '../Styles'

import FormError from '../FormError'
import FormSuccess from '../FormSuccess'
import Button, { VARIANTS } from '../Button'
import ButtonGroup from '../Button/Group'
import SectionTitle from '../SectionTitle'

const ProjectUsersAddFormResponse = ({
  projectId,
  succeeded,
  failed,
  permissions: p,
  onReset,
}) => {
  const permissions = Object.keys(p).filter(name => p[name])

  return (
    <div>
      {failed.length > 0 && (
        <>
          <FormError>Users Not Added!</FormError>
          <SectionTitle>Users that Need an Account</SectionTitle>
          <div
            css={{
              paddingTop: spacing.moderate,
              paddingBottom: spacing.moderate,
              color: colors.structure.steel,
            }}>
            {failed.map(user => user.email).join(', ')}
          </div>
        </>
      )}

      {succeeded.length > 0 && (
        <>
          <FormSuccess>Users Added!</FormSuccess>

          <SectionTitle>Users Added</SectionTitle>
          <div
            css={{
              paddingTop: spacing.moderate,
              color: colors.structure.steel,
            }}>
            {succeeded.map(user => user.email).join(', ')}
          </div>

          <SectionTitle>Permissions</SectionTitle>
          <ul css={{ color: colors.structure.zinc }}>
            {permissions.map(permission => (
              <li css={{ fontWeight: typography.weight.bold }} key={permission}>
                {permission.replace(/([A-Z])/g, match => ` ${match}`)}
              </li>
            ))}
          </ul>

          <ButtonGroup>
            <Button variant={VARIANTS.SECONDARY} onClick={onReset}>
              Add Another User
            </Button>
            <Link href="/[projectId]/users" as={`/${projectId}/users`} passHref>
              <Button variant={VARIANTS.PRIMARY}>View All</Button>
            </Link>
          </ButtonGroup>
        </>
      )}
    </div>
  )
}

ProjectUsersAddFormResponse.propTypes = {
  projectId: PropTypes.string.isRequired,
  succeeded: PropTypes.arrayOf(
    PropTypes.shape({
      email: PropTypes.string.isRequired,
      permissions: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    }).isRequired,
  ).isRequired,
  failed: PropTypes.arrayOf(
    PropTypes.shape({
      email: PropTypes.string.isRequired,
      permissions: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    }).isRequired,
  ).isRequired,
  permissions: PropTypes.objectOf(PropTypes.bool.isRequired).isRequired,
  onReset: PropTypes.func.isRequired,
}

export default ProjectUsersAddFormResponse
