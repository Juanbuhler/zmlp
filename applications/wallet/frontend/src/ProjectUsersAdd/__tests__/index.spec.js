import TestRenderer, { act } from 'react-test-renderer'

import permissions from '../../Permissions/__mocks__/permissions'

import ProjectUsersAdd from '..'

const PROJECT_ID = '76917058-b147-4556-987a-0a0f11e46d9b'

jest.mock('../helpers')

const noop = () => () => {}

describe('<ProjectUsersAdd />', () => {
  it('should render properly', () => {
    require('next/router').__setUseRouter({
      pathname: '/[projectId]/users/add',
      query: { projectId: PROJECT_ID },
    })

    require('swr').__setMockUseSWRResponse({
      data: permissions,
    })

    const mockOnCopy = jest.fn()

    require('../helpers').__setMockOnCopy(mockOnCopy)

    const component = TestRenderer.create(<ProjectUsersAdd />)

    expect(component.toJSON()).toMatchSnapshot()

    // Input email
    act(() => {
      component.root
        .findByProps({ name: 'emails' })
        .props.onChange({ target: { value: 'jane@email.com' } })
    })

    // Check permission box
    act(() => {
      component.root
        .findByProps({ type: 'checkbox', value: 'SystemProjectOverride' })
        .props.onClick()
    })

    expect(component.toJSON()).toMatchSnapshot()

    // Submit the form
    act(() => {
      component.root
        .findByProps({ children: 'Add' })
        .props.onClick({ preventDefault: noop })
    })

    expect(component.toJSON()).toMatchSnapshot()

    // Copy Key to clipboard
    act(() => {
      component.root
        .findByProps({ children: 'Copy Link' })
        .props.onClick({ preventDefault: noop })
    })

    expect(mockOnCopy).toHaveBeenCalledWith({ inputRef: { current: null } })

    // Reset form
    act(() => {
      component.root
        .findByProps({ children: 'Add Another User' })
        .props.onClick()
    })

    expect(component.toJSON()).toMatchSnapshot()
  })
})
