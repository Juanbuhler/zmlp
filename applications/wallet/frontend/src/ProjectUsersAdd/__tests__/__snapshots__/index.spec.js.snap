// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectUsersAdd /> should render properly 1`] = `
Array [
  .emotion-0 {
  font-size: 18px;
  line-height: 21px;
  font-weight: 400;
  padding-top: 24px;
  padding-bottom: 16px;
}

<h2
    className="emotion-0"
  >
    Project User Admin
  </h2>,
  .emotion-5 {
  padding: 16px;
  padding-left: 0;
  padding-right: 0;
}

.emotion-4 {
  list-style-type: none;
  padding: 0;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1 {
  padding-right: 16px;
}

.emotion-0 {
  border: 0 #b8d652 solid;
  border-bottom-width: 0;
  color: #b3b3b3;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-top: 8px;
  padding-bottom: 8px;
  text-transform: uppercase;
  font-size: 17px;
  line-height: 20px;
  font-weight: 700;
}

.emotion-0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #73b61c;
}

<nav
    className="emotion-5"
  >
    <ul
      className="emotion-4"
    >
      <li
        className="emotion-1"
      >
        <a
          className="emotion-0"
          href="/76917058-b147-4556-987a-0a0f11e46d9b/users"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Users
        </a>
      </li>
      <li
        className="emotion-1"
      >
        <a
          className="emotion-0"
          href="/76917058-b147-4556-987a-0a0f11e46d9b/users/add"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Create User
        </a>
      </li>
    </ul>
  </nav>,
  <div>
    Â 
  </div>,
]
`;
